source/common/http/http2/codec_impl.cc

```diff
@@ -504,14 +510,26 @@ ssize_t ConnectionImpl::onSend(const uint8_t* data, size_t length) {
 }
 
 int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
-
   StreamImpl* stream = getStream(stream_id);
   if (stream) {
     ENVOY_CONN_LOG(debug, "stream closed: {}", connection_, error_code);
     if (!stream->remote_end_stream_ || !stream->local_end_stream_) {
-      stream->runResetCallbacks(error_code == NGHTTP2_REFUSED_STREAM
-                                    ? StreamResetReason::RemoteRefusedStreamReset
-                                    : StreamResetReason::RemoteReset);
+      StreamResetReason reason;
+      if (stream->reset_due_to_messaging_error_) {
```

Comment:
https://nghttp2.org/documentation/types.html#c.nghttp2_on_stream_close_callback indicates that error_code is *generally* nghttp2_error_code here.   
Are there cases where we'll get onInvalidFrame(NGHTTP2_ERR_HTTP_MESSAGING) and onStreamClose(SomeOtherError)?  I'm not terribly familiar with the internals of NGHTTP2_ERR_HTTP_MESSAGING but I wouldn't have assumed all messaging errors triggered invalidFrame so I would assume setting reset_due_to_messaging_error_  where we reset the stream would more often be correct.

Reply:
Unfortunately this is the issue that I was commenting on re: gymnastics of doing this change. By the time we get to onStreamClose(), we only get one of these errors: https://nghttp2.org/documentation/enums.html#c.nghttp2_error_code which are RFC wire errors. So all lib errors (https://nghttp2.org/documentation/enums.html#c.nghttp2_error) have been collapsed into a generic error. It would be nice if nghttp2 also gave us the lib error that caused the wire error but I don't think there is any way to get that. What I did here is I think the "cleanest" way of moving from lib error to reset reason. AFAIK, all messaging errors will trigger invalid frame, because a messaging error came from some frame (headers, data, etc.) so that frame will be rejected before it is dispatched.

@tatsuhiro-t in case you have a few moments to review this change, is there any cleaner way of accomplishing what I'm trying to do?

Reply:
Argh, I was conflating nghttp2_error and nghttp2_error_code

Totally PEBKAC but I'm going to claim it's an understandable mistake :-P

Reply:
I think this PR is best thing I can think of if you want to treat HTTP messaging error differently.